servermaster/job_fsm.go

```diff
@@ -22,32 +22,45 @@ type jobHolder struct {
-//     |                           |                    |
-//     |                           | Master             |
-//     |                           |   .OnWorkerOffline |
-//     |                           |------------------->|
-//     |                           |                    |
-//     |                           |                    |
-//     |                           | Master             |
-//     |                           |   .CreateWorker    |
-//     |<-----------------------------------------------|
-//     |                           |                    |
-//     |                           |                    |
-//     | Master                    |                    |
-//     |  .OnWorkerDispatched      |                    |
-//     |  (with error)             |                    |
-//     |----------------------------------------------->|
-//     |                           |                    |
-//     |                           |                    |
-//     |                           |                    |
+// ,-------.                   ,-------.            ,-------.       ,--------.
+// |WaitAck|                   |Online |            |Pending|       |Finished|
+// `---+---'                   `---+---'            `---+---'       `---+----'
+//     |                           |                    |               |
+//     | Master                    |                    |               |
+//     |  .OnWorkerOnline          |                    |               |
+//     |-------------------------->|                    |               |
+//     |                           |                    |               |
+//     |                           | Master             |               |
+//     |                           |   .OnWorkerOffline |               |
+//     |                           |   (failover)       |               |
+//     |                           |------------------->|               |
+//     |                           |                    |               |
+//     |                           | Master             |               |
+//     |                           |   .OnWorkerOffline |               |
+//     |                           |   (finish)         |               |
+//     |                           |----------------------------------->|
+//     |                           |                    |               |
+//     | Master                    |                    |               |
+//     |  .OnWorkerOffline         |                    |               |
+//     |  (failover)               |                    |               |
+//     |----------------------------------------------->|               |
+//     |                           |                    |               |
+//     | Master                    |                    |               |
+//     |  .OnWorkerOffline         |                    |               |
```

Comment:
Can the worker check the first heartbeat pong before it do some work?
Seems strange that the worker state is marked finish before it's online.

Reply:
They are separated logic, the fsm can't control whether heartbeat is received.
Besides the worker could finish before any heartbeat is sent.

Reply:
this problem should be solved by the worker

Reply:
Worker can try to send one heartbeat during initialization, but we can't ensure the pong message is received, which can infer `OnWorkerOnline` is called in master, this is caused by an underlying assumption: **Worker should continue to work without confirming the heartbeat between master is normal.**
Supposing the following scenario
- During worker initialization, the heartbeat ping message is blocking or heartbeat ping is sent, but heartbeat pong is not received, should the worker continues to work now?

Reply:
I think the worker should continue work